src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/InteropCrypto.cs

```diff
@@ -0,0 +1,165 @@
+
+        private static class WindowsLib
+        {
+            public const string CryptoDll = "libeay32.dll";
+            public const string SslDll = "ssleay32.dll";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void CRYPTO_set_locking_callback(locking_function lockingFunction);
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void OPENSSL_add_all_algorithms_noconf();
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void ERR_load_crypto_strings();
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void CRYPTO_free(void* pointer);
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static int CRYPTO_num_locks();
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void SSL_load_error_strings();
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static int SSL_library_init();
+
+            public static void Init()
+            {
+                CRYPTO_set_locking_callback(LockStore.Callback);
+                ERR_load_crypto_strings();
+                SSL_load_error_strings();
+                OPENSSL_add_all_algorithms_noconf();
+                CheckForErrorOrThrow(SSL_library_init());
+            }
+        }
+
+        private static class UnixLib
+        {
+            public const string CryptoDll = "libcrypto.so";
+            public const string SslDll = "libssl.so";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void CRYPTO_set_locking_callback(locking_function lockingFunction);
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void OPENSSL_add_all_algorithms_noconf();
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void ERR_load_crypto_strings();
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void CRYPTO_free(void* pointer);
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static int CRYPTO_num_locks();
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void SSL_load_error_strings();
+            [DllImport(SslDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static int SSL_library_init();
+
+            public static void Init()
+            {
+                CRYPTO_set_locking_callback(LockStore.Callback);
```

Comment:
This may dangerously interact with anything in-proc already using OpenSSL, such as the current crypto stack.